File: NeetCode/Sliding_Window/Find_K_Closest_Elements/find__k__closest__elements.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        left = 0
        right = len(arr) - k

        while left < right:
            mid = left + (right - left) // 2

            if x - arr[mid] > arr[mid + k] - x:
                left = mid + 1
            else:
                right = mid

        return arr[left : left + k]


if __name__ == "__main__":
    sol = Solution()
    arr = [2, 4, 5, 8]
    k = 2
    x = 6
    print(sol.findClosestElements(arr, k, x))
    print("Running Solution...")

File: NeetCode/Sliding_Window/Find_K_Closest_Elements/Find_K_Closest_Elements.cpp
// Find_K_Closest_Elements.cpp
//
// Compile locally with:
// clang++ Find_K_Closest_Elements.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<int> findClosestElements(vector<int> &arr, int k, int x)
    {
        int left(0), mid(0);
        int right = arr.size() - k;

        while (left < right)
        {
            mid = left + (right - left) / 2;

            if (x - arr[mid] > arr[mid + k] - x)
            {
                left = mid + 1;
            }
            else
            {
                right = mid;
            }
        }
        return vector<int>(arr.begin() + left, arr.begin() + left + k);
    }
};

int main()
{
    cout << "Running Find_K_Closest_Elements.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
